### PR TITLE
Prototype for monitor_dbus_signals test

### PR DESCRIPTION
**Pull request by bgurney-rh**
Supersedes: #67
Related: https://github.com/stratis-storage/project/issues/369

### CODE REVIEW

**Comment by bgurney-rh** [comment]
On a test run of stratisd_optional.py, the tests pass, but it looks like the trace process is cleaned up too late:
```
test_filesystem_create (__main__.StratisdCertify)
Test creating a filesystem. ... /usr/lib64/python3.9/subprocess.py:1052: ResourceWarning: subprocess 23406 is still running
  _warn("subprocess %s is still running" % self.pid,
ResourceWarning: Enable tracemalloc to get the object allocation traceback
ok
```

File: testlib/utils.py [hunk 1]
```diff
@@ -76,6 +76,29 @@ def process_exists(name):
...
+    intended for Python scripts whose name will be in the cmdline, but
+    not in the process name.
+    :param name: name of script to clean up
+    :type name: str
+    return: None
+    """
+    procs = []
+    for proc in psutil.process_iter(["cmdline"]):
+        try:
+            for param in proc.info["cmdline"]:
```
Comment (mulkieran) [inline testlib/utils.py:91]: Here's a place where you could get fancy with list comprehension and this time probably increase comprehensibility:

```
if any(name in param for param in proc.info["cmdline"]):
   procs.append(proc)
```

Of course, as we discussed, it probably makes sense just to kill the process as soon as you find it...

**Comment by mulkieran** [comment]
Let's not worry about the ResourceWarnings until we're entirely done w/ development. I think that once we've completed all the changes, they will go away.

File: testlib/utils.py (hunk 1 above)
Comment (mulkieran) [reply testlib/utils.py:91]: Also, I think I'ld prefer ``` name == param ``` to ```name in param```. If the name argument is "a", this function could end up killing a lot of processes.

Comment (bgurney-rh) [reply testlib/utils.py]: Resolved.

**Comment by bgurney-rh** [comment]
Now I need to figure out how to call `trace.send_signal(signal.SIGINT)` in one of the tests.  I might have to change something in the `setUp()` method to be able to access `trace` from a point during a test.

File: stratisd_optional.py [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+            time.sleep(20)
+
+        clean_up()
+
+        time.sleep(1)
+        exec_command(["udevadm", "settle"])
+
+        trace = subprocess.Popen(
+            [
+                "./scripts/monitor_dbus_signals.py",
```
Comment (mulkieran) [inline stratisd_optional.py:203]: We should create a constant for this string, since we use it in two places.

**Review by mulkieran — CHANGES_REQUESTED** [review]

File: stratisd_optional.py (hunk 2 above)
Comment (bgurney-rh) [reply stratisd_optional.py]: Resolved.

**Comment by bgurney-rh** [comment]
Right now, the test fails if the monitor_dbus_signals test has a non-zero return code. However, I need to find a way to get the stderr from the script into the test failure traceback (instead of appearing in stderr during the test run).

**Comment by bgurney-rh** [comment]
Rebased.

**Comment by mulkieran** [comment]
@bgurney-rh Looks good so far. I think we decided that the next step should be to make whether or not the script is run be an option specified at the command-line. The default should be not to run the script. So, you'll want to add an additional boolean option, like ```--monitor-dbus``` and make sure it is respected.

**Comment by bgurney-rh** [comment]
OK, I added the `--monitor-dbus` option; if it's not provided, the program prints the help text.  (And please let me know if I should add anything for the help text of `--monitor-dbus`.)

**Comment by bgurney-rh** [comment]
Rebased.

**Comment by mulkieran** [comment]
@bgurney-rh It makes more sense to specify whether or not the D-Bus is monitored via an option to stratisd_cert.py, not via an option to the D-Bus monitoring script. If the decision is not to monitor the D-Bus then it is much better for stratisd_cert.py not to start the separate process for the script at all.

**Comment by bgurney-rh** [comment]
I created a "v2" pull request to perform the monitor_dbus_signals test in the stratisd_cert test.  The test run takes about 5 minutes, and I saw 14 tests fail with `--monitor-dbus` enabled; some of those failures may be things we haven't seen before.

**Comment by mulkieran** [comment]
@bgurney-rh I believe you can close this one.

**Pull request closed without merging**